modules/nextflow/src/main/groovy/nextflow/file/FilePorter.groovy

```diff
@@ -76,32 +77,31 @@ class FilePorter {
                 .createAndRegisterShutdownCallback(session)
     }
 
-    Batch newBatch(Path stageDir) { new Batch(stageDir) }
+    Batch newBatch(Path stageDir) { new Batch(this, stageDir) }
 
     void transfer(Batch batch) {
         if( batch.size() ) {
             log.trace "Stage foreign files: $batch"
-            submitStagingActions(batch.foreignPaths, batch.stageDir)
+            submitStagingActions(batch.foreignPaths)
             log.trace "Stage foreign files completed: $batch"
         }
     }
 
-    protected FileTransfer createFileTransfer(Path source, Path stageDir) {
-        final stagePath = getCachePathFor(source,stageDir)
```

Comment:
This is the main change. Instead of recomputing it, use the target path that was computed in the previous step